README.md
Custom Admin Login
===========
Allows you to customize the background, logo, font color, url and caption on the WordPress login page.

Features
-----------
You can:
* Set the color and/or image background.
* Set the logo image contained in a area 320 x 84 pixels (if it's bigger, it's scaled to fit).
* Set URL and caption of the logo image.
* Set font color outside of login form.

Bug tracker
-----------
Have a bug? Please create an issue on GitHub [here](https://github.com/jprieton/custom-admin-login/issues)


CHANGELOG.md
# CHANGELOG

Custom Admin Login follows [SemVer](http://semver.org/).

## 1.x

- 1.0.0
  - Initial release

- 1.0.4
  - Update deprecated filters

custom-admin-login.php
<?php

/**
 * Plugin Name:         Custom Admin Login
 * Plugin URI:          https://github.com/jprieton/custom-admin-login
 * Description:         Allows you to customize the background, logo, url and caption on the WordPress login page.
 * Tags:                admin, login, custom, background, logo, custom admin login, login page
 * Version:             1.0.4
 * Requires at least:   5.2
 * Tested up to:        5.2.1
 * Author:              Javier Prieto
 * Author URI:          https://github.com/jprieton
 * Text Domain:         custom-admin-login
 * Domain Path:         /languages/
 *
 * Custom Admin Login is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * Custom Admin Login is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Custom Admin Login. If not, see http://www.gnu.org/licenses/gpl-3.0.txt.
 *
 * @package CustomAdminLogin
 */

// If this file is called directly, abort.
defined( 'ABSPATH' ) || exit;

/**
 * Define plugin constants
 * @since 1.0.0
 */
define( 'CustomAdminLogin\FILENAME', __FILE__ );
define( 'CustomAdminLogin\BASENAME', plugin_basename( __FILE__ ) );
define( 'CustomAdminLogin\INCLUDES', plugin_dir_path( CustomAdminLogin\FILENAME ) . 'includes' );

// Load texdomain
require_once CustomAdminLogin\INCLUDES . '/class-textdomain.php';
CustomAdminLogin\Textdomain::init();

// Initialize customizer
require_once CustomAdminLogin\INCLUDES . '/class-theme-customizer.php';
CustomAdminLogin\Theme_Customizer::init();



admin-login-customizer.php
<?php

/**
 * Plugin Name:         Admin Login Customizer
 * Plugin URI:          https://github.com/jprieton/admin-login-customizer
 * Description:         Allows you to customize the background, logo, url and caption on the WordPress login page.
 * Tags:                admin, login, custom
 * Version:             1.0.0
 * Requires at least:   4.9
 * Tested up to:        4.9.6
 * Author:              Javier Prieto
 * Author URI:          https://github.com/jprieton
 * Text Domain:         admin-login-customizer
 * Domain Path:         /languages/
 *
 * Admin Login Customizer is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * any later version.
 *
 * Admin Login Customizer is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Admin Login Customizer. If not, see http://www.gnu.org/licenses/gpl-3.0.txt.
 *
 * @package AdminLoginCustomizer
 */
// If this file is called directly, abort.
if ( !defined( 'ABSPATH' ) ) {
  die( 'Direct access is forbidden.' );
}

/**
 * Define plugin constants
 * @since 1.0.0
 */
define( 'ALC_VERSION', '1.0.0' );
define( 'ALC_FILENAME', __FILE__ );
define( 'ALC_BASENAME', plugin_basename( __FILE__ ) );
define( 'ALC_BASEDIR', __DIR__ );
define( 'ALC_BASEURL', plugin_dir_url( __FILE__ ) );
define( 'ALC_TEXTDOMAIN', 'admin-login-customizer' );
define( 'ALC_ABSPATH', plugin_dir_path( ALC_FILENAME ) . 'includes' );

// Load texdomain
require_once ALC_ABSPATH . '/class-textdomain.php';
ALC\Textdomain::init();

// Initialize customizer
require_once ALC_ABSPATH . '/class-theme-customizer.php';
ALC\Theme_Customizer::init();



includes/class-textdomain.php
<?php

namespace CustomAdminLogin;

// If this file is called directly, abort.
defined( 'ABSPATH' ) || exit;

/**
 * Textdomain class
 *
 * @package        AdminLoginCustomizer
 * @subpackage     Init
 * @since          1.0.0
 * @author         Javier Prieto
 */
class Textdomain {

  /**
   * Single instance of this class
   *
   * @since     1.0.0
   * @var       Theme_Customizer
   */
  protected static $instance;

  /**
   * Main instance
   * Ensures only one instance of this class is loaded or can be loaded.
   *
   * @since   1.0.0
   * @static
   * @return  static
   */
  public static function &init() {
    if ( empty( static::$instance ) ) {
      static::$instance = new static();
    }

    return static::$instance;
  }

  /**
   * Class constructor
   * Declared as protected to prevent creating a new instance outside of the class via the new operator.
   *
   * @since   1.0.0
   */
  protected function __construct() {
    // This hook load the plugin textdomain
    add_action( 'plugins_loaded', [ $this, 'load_plugin_textdomain' ] );

    // This hook adds tanslation to plugin description
    add_action( 'all_plugins', [ $this, 'modify_plugin_description' ] );
  }

  /**
   * Load plugin textdomain
   *
   * @since     1.1.0
   */
  public static function load_plugin_textdomain() {
    load_plugin_textdomain( 'custom-admin-login', FALSE, basename( dirname( BASENAME ) ) . '/languages/' );
  }

  /**
   * Adds tanslation to plugin description
   *
   * @since     1.0.0
   * @param     array $all_plugins
   * @return    array
   */
  public static function modify_plugin_description( $all_plugins = [] ) {
    if ( key_exists( BASENAME, $all_plugins ) ) {
      $all_plugins[BASENAME]['Description'] = __( 'Allows you to customize the background, logo, url and caption on the WordPress login page.', 'custom-admin-login' );
    }
    return $all_plugins;
  }

}


includes/class-theme-customizer.php
<?php

namespace CustomAdminLogin;

// If this file is called directly, abort.
defined( 'ABSPATH' ) || exit;

use WP_Customize_Manager;
use WP_Customize_Color_Control;
use WP_Customize_Image_Control;

/**
 * Theme_Customizer class
 *
 * @package        AdminLoginCustomizer
 * @subpackage     Admin
 * @since          1.0.0
 * @author         Javier Prieto
 */
class Theme_Customizer {

  /**
   * Single instance of this class
   *
   * @since     1.0.0
   * @var       Theme_Customizer
   */
  protected static $instance;

  /**
   * Main instance
   * Ensures only one instance of this class is loaded or can be loaded.
   *
   * @since   1.0.0
   * @static
   * @return  static
   */
  public static function &init() {
    if ( empty( self::$instance ) ) {
      self::$instance = new static();
    }

    return self::$instance;
  }

  /**
   * Class constructor
   * Declared as protected to prevent creating a new instance outside of the class via the new operator.
   *
   * @since   1.0.0
   */
  protected function __construct() {
    // admin customizer
    add_action( 'customize_register', [ $this, 'login_page' ] );

    // login page
    add_filter( 'login_headerurl', [ $this, 'login_header_url' ], 99 );
    add_filter( 'login_headertext', [ $this, 'login_header_title' ], 99 );
    add_action( 'login_enqueue_scripts', [ $this, 'login_header_image' ], 99 );
    add_action( 'login_enqueue_scripts', [ $this, 'login_font_color' ], 99 );
    add_action( 'login_enqueue_scripts', [ $this, 'login_background_color' ], 99 );
    add_action( 'login_enqueue_scripts', [ $this, 'login_background' ], 99 );
  }

  /**
   * Adds the login page menu to theme customizer
   *
   * @since   1.0.0
   *
   * @param WP_Customize_Manager $wp_customize
   */
  public function login_page( $wp_customize ) {
    $section_id = 'login_customizer_section';

    $wp_customize->add_section( $section_id, [
        'title'    => __( 'Login Page', 'custom-admin-login' ),
        'priority' => 1000,
    ] );

    $wp_customize->add_setting( 'login_background_color', [ 'default' => '#f1f1f1', ] );
    $wp_customize->add_control(
            new WP_Customize_Color_Control( $wp_customize, 'login_background_color', [
                'label'    => __( 'Background color', 'custom-admin-login' ),
                'section'  => $section_id,
                'settings' => 'login_background_color',
                    ] )
    );

    $wp_customize->add_setting( 'login_background_image' );
    $wp_customize->add_control(
            new WP_Customize_Image_Control( $wp_customize, 'login_background_image', [
                'label'    => __( 'Background image', 'custom-admin-login' ),
                'section'  => $section_id,
                'settings' => 'login_background_image',
                    ] )
    );

    $wp_customize->add_setting( 'login_background_position' );
    $wp_customize->add_control( 'login_background_position', [
        'label'    => __( 'Background position', 'custom-admin-login' ),
        'section'  => $section_id,
        'settings' => 'login_background_position',
        'type'     => 'select',
        'choices'  => [
            ''        => 'Auto',
            'contain' => 'Contain',
            'cover'   => 'Cover',
            'repeat'  => 'Repeat',
        ]
    ] );

    $wp_customize->add_setting( 'login_font_color', [ 'default' => '#555d66', ] );
    $wp_customize->add_control(
            new WP_Customize_Color_Control( $wp_customize, 'login_font_color', [
                'label'    => __( 'Font color', 'custom-admin-login' ),
                'section'  => $section_id,
                'settings' => 'login_font_color',
                    ] )
    );

    $wp_customize->add_setting( 'login_header_image' );
    $wp_customize->add_control(
            new WP_Customize_Image_Control( $wp_customize, 'login_header_image', [
                'label'       => __( 'Header image', 'custom-admin-login' ),
                'section'     => $section_id,
                'settings'    => 'login_header_image',
                'description' => __( 'The custom header is centered and contained in a 320 x 84 pixels block', 'custom-admin-login' ),
                    ] )
    );

    $wp_customize->add_setting( 'login_header_url' );
    $wp_customize->add_control( 'login_header_url', [
        'label'    => __( 'Header URL', 'custom-admin-login' ),
        'section'  => $section_id,
        'settings' => 'login_header_url',
        'type'     => 'text',
    ] );

    $wp_customize->add_setting( 'login_header_title' );
    $wp_customize->add_control( 'login_header_title', [
        'label'    => __( 'Header Title', 'custom-admin-login' ),
        'section'  => $section_id,
        'settings' => 'login_header_title',
        'type'     => 'text',
    ] );
  }

  /**
   * Replace the header url with custom url
   *
   * @since   1.0.0
   *
   * @param   string $url
   * @return  string
   */
  public function login_header_url( $url ) {
    $custom_url = get_theme_mod( 'login_header_url' );
    return $custom_url ?: $url;
  }

  /**
   * Replace the header title with custom string
   *
   * @since   1.0.0
   *
   * @param   string $title
   * @return  string
   */
  public function login_header_title( $title ) {
    $custom_title = get_theme_mod( 'login_header_title' );
    return $custom_title ?: $title;
  }

  /**
   * Replace the header WordPress logo with custom image
   *
   * @since   1.0.0
   *
   * @return  string
   */
  public function login_header_image() {
    $custom_image = get_theme_mod( 'login_header_image' );

    if ( !empty( $custom_image ) ) {
      ?>
      <style type="text/css">
        #login h1 a {
          background-image: url(<?php echo $custom_image ?>);
          background-position: center;
          background-size: contain;
          height: 84px;
          width: 320px;
        }
      </style>
      <?php
    }
  }

  /**
   * Replace the font color of WordPress login page
   *
   * @since   1.0.0
   *
   * @return  string
   */
  public function login_font_color() {
    $custom_font_color = get_theme_mod( 'login_font_color' );

    if ( !empty( $custom_font_color ) ) {
      ?>
      <style type="text/css">
        body.login #backtoblog a, body.login #nav a {
          color: <?php echo $custom_font_color ?>;
        }
      </style>
      <?php
    }
  }

  /**
   * Replace the background color of WordPress login page
   *
   * @since   1.0.0
   *
   * @return  string
   */
  public function login_background_color() {
    $custom_background_color = get_theme_mod( 'login_background_color' );

    if ( !empty( $custom_background_color ) ) {
      ?>
      <style type="text/css">
        body.login  {
          background-color: <?php echo $custom_background_color ?>;
        }
      </style>
      <?php
    }
  }

  /**
   * Sets the background position of WordPress login page
   *
   * @since   1.0.0
   *
   * @return  string
   */
  public function login_background() {
    $background_position = get_theme_mod( 'login_background_position' );
    $background_image    = get_theme_mod( 'login_background_image' );

    $position = 'center';
    $size     = 'auto';
    $repeat   = 'no-repeat';

    switch ( $background_position ) {
      case 'contain':
        $size = 'contain';
        break;

      case 'cover':
        $size = 'cover';
        break;

      case 'repeat':
        $repeat = 'repeat';
        break;
    }

    if ( !empty( $background_image ) ) {
      ?>
      <style type="text/css">
        body.login  {
          background-image: url(<?php echo $background_image ?>);
          background-position: <?php echo $position ?>;
          background-size:  <?php echo $size ?>;
          background-repeat:  <?php echo $repeat ?>;
        }
      </style>
      <?php
    }
  }

  /**
   * Declared as private to prevent cloning of an instance of the class via the clone operator.
   *
   * @since   1.0.0
   */
  private function __clone() {
    
  }

  /**
   * Declared as private to prevent unserializing of an instance of the class via the global function unserialize().
   *
   * @since   1.0.0
   */
  private function __wakeup() {
    
  }

  /**
   * Declared as protected to prevent serializg of an instance of the class via the global function serialize().
   *
   * @since   1.0.0
   */
  protected function __sleep() {
    
  }

}
